From eea69204371883329affad07879d98d0629c01cd Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Fri, 11 Jul 2025 15:40:52 +0800
Subject: [PATCH 106/123] mtk: mt76: mt7996: add dfs channel check for
 triggering radar detection

In non-detection mode, when radar detection is triggered, the firmware
automatically stops the TX queue.
In our design, the TX queue is resumed after a channel switch.
Therefore, triggering radar detection on a non-DFS channel will cause
the TX queue to remain blocked and not resume until the interface is reloaded,
since a channel switch is not required in this case.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt7996/debugfs.c | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/mt7996/debugfs.c b/mt7996/debugfs.c
index ee935928..fba1b828 100644
--- a/mt7996/debugfs.c
+++ b/mt7996/debugfs.c
@@ -262,7 +262,11 @@ mt7996_radar_trigger(void *data, u64 val)
 #define RADAR_BACKGROUND	2
 	struct mt7996_phy *phy = data;
 	struct mt7996_dev *dev = phy->dev;
-	int rdd_idx;
+	struct cfg80211_chan_def *chandef = &phy->mt76->chandef;
+	int rdd_idx, ret;
+
+	if (test_bit(MT76_SCANNING, &phy->mt76->state))
+		return -EBUSY;
 
 	if (!val || val > RADAR_BACKGROUND)
 		return -EINVAL;
@@ -278,6 +282,17 @@ mt7996_radar_trigger(void *data, u64 val)
 		return -EINVAL;
 	}
 
+	ret = cfg80211_chandef_dfs_required(dev->mt76.hw->wiphy, chandef, NL80211_IFTYPE_AP);
+	if (ret <= 0) {
+		if (ret < 0)
+			return ret;
+
+		dev_err(dev->mt76.dev,
+			"Cannot trigger radar detection on non-DFS channel %d width %d\n",
+			chandef->chan->hw_value, chandef->width);
+		return -EINVAL;
+	}
+
 	return mt7996_mcu_rdd_cmd(dev, RDD_RADAR_EMULATE, rdd_idx, 0);
 }
 
-- 
2.45.2

